dali/kernels/slice/slice_flip_normalize_permute_pad_cpu.h

```diff
@@ -30,165 +31,291 @@ namespace kernels {
-                          const float *mean, const float *inv_stddev) {
-    float fpout = (static_cast<float>(element) - (*mean)) * (*inv_stddev);
-    if (std::is_integral<OutputType>::value) {
-      destination = clamp<OutputType>(std::roundf(fpout));
+    // out of bounds (left side)
+    if (HasChannels && d == channel_dim) {
+      for (int64_t i = 0; i < pad_before; i++) {
+        *output = *fill_values;
+        output += 1;  // out_strides[d] is 1;
+        input  += in_strides[d];
+        fill_values++;
+        mean++;
+        inv_stddev++;
+      }
     } else {
-      destination = clamp<OutputType>(fpout);
+      for (int64_t i = 0; i < pad_before; i++) {
+        *output = *fill_values;
+        output += 1;  // out_strides[d] is 1;
+        input  += in_strides[d];
+      }
     }
-  }
-};
 
-template <typename OutputType, int Dims>
-inline void Pad(OutputType *output,
-                TensorShape<Dims> out_strides,
-                TensorShape<Dims> padded_out_shape,
-                float fill_value,
-                std::integral_constant<int, 1>) {
-  constexpr auto d = Dims - 1;
-  for (int64_t i = 0; i < padded_out_shape[d]; i++) {
-    *output = Convert<OutputType>(fill_value);
-    output += out_strides[d];
+    // within input bounds
+    if (HasChannels && d == channel_dim) {
+      for (int64_t i = 0; i < slice; i++) {
+        Fill<NeedNormalize>(*output, *input, mean, inv_stddev);
+        output += 1;  // out_strides[d] is 1;
+        input  += in_strides[d];
+        fill_values++;
+        mean++;
+        inv_stddev++;
+      }
+    } else {
+      for (int64_t i = 0; i < slice; i++) {
+        Fill<NeedNormalize>(*output, *input, mean, inv_stddev);
+        output += 1;  // out_strides[d] is 1;
+        input  += in_strides[d];
+      }
+    }
+
+    // out of bounds (right side)
+    if (HasChannels && d == channel_dim) {
+      for (int64_t i = 0; i < pad_after; i++) {
+        *output = *fill_values;
+        output += 1;  // out_strides[d] is 1;
+        input += in_strides[d];
+        fill_values++;
+        mean++;
+        inv_stddev++;
+      }
+    } else {
+      for (int64_t i = 0; i < pad_after; i++) {
+        *output = *fill_values;
+        output += 1;  // out_strides[d] is 1;
+        input += in_strides[d];
+      }
+    }
   }
 }
 
-template <typename OutputType, int Dims, int DimsLeft>
-inline void Pad(OutputType *output,
-                TensorShape<Dims> out_strides,
-                TensorShape<Dims> padded_out_shape,
-                float fill_value,
-                std::integral_constant<int, DimsLeft>) {
-  constexpr auto d = Dims - DimsLeft;
-  for (int64_t i = 0; i < padded_out_shape[d]; i++) {
-    Pad(output, out_strides, padded_out_shape, fill_value,
-        std::integral_constant<int, DimsLeft - 1>());
-    output += out_strides[d];
+template <bool NeedNormalize, bool HasChannels, typename OutputType, typename InputType>
+void SliceFlipNormalizePermuteKernelImpl(
```

Comment:
I would move that after `SliceFlipNormalizePermutePadKernelImpl` bellow. To have all variants of `SliceFlipNormalizePermutePadKernelImpl` together.